### PR TITLE
Provisioning support for ZooKeeper Authorization

### PR DESCRIPTION
**Pull request by varaga**
The patch fix enables Druid nodes to authenticate with ZooKeeper using username, paasword, auth schemes as supported by the ZooKeeper.

The issue is discussed and addressed through discussion here: https://github.com/druid-io/druid/issues/5641

### CODE REVIEW

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 1]
```diff
@@ -67,18 +71,21 @@ public void configure(Binder binder)
...
-                               .sessionTimeoutMs(config.getZkSessionTimeoutMs())
-                               .retryPolicy(new BoundedExponentialBackoffRetry(
-                                   BASE_SLEEP_TIME_MS, MAX_SLEEP_TIME_MS, MAX_RETRIES))
-                               .compressionProvider(new PotentiallyGzippedCompressionProvider(config.getEnableCompression()))
-                               .aclProvider(config.getEnableAcl() ? new SecuredACLProvider() : new DefaultACLProvider())
-                               .build();
+    final Builder builder = CuratorFrameworkFactory.builder();
+    if (null != config.getZkUser() && null != config.getZkPwd()) {
+      builder.authorization(config.getAuthScheme(),
+          (config.getZkUser() + ":" + config.getZkPwd()).getBytes(StandardCharsets.UTF_8));
```
Comment (b-slim) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: am wondering is there any specs about how to go from string to bytes?

**Comment by b-slim** [comment]
@varaga is it possible to unit test this with good/bad passwords ?

File: server/src/main/java/io/druid/curator/CuratorModule.java (hunk 1 above)
Comment (varaga) [reply server/src/main/java/io/druid/curator/CuratorModule.java]: @b-slim - There isn't a spec., provided by ZK for charsets. Essentially, all fetches from ZK Server by the clients are base64 encoded. From the client-side using UTF-8 is more consistent across platforms than relying on platform's charset.
BTW., I tested it fine against ZK cluster running on AWS.

**Comment by varaga** [comment]
@b-slim - The expected validation exception is buried deep in the stack and is returned from ZK server. I don't see value in mocking the stack up to verify the semantics of a wrong password. Secondly, I don't see ZK integration tests in druid that verifies ZK config, likes of acl, paths etc., against the ZK server. What do you reckon?

**Comment by b-slim** [comment]
>  I don't see value in mocking the stack up to verify the semantics of a wrong password. 

Well there is multiple values one of them is to catch bugs if someone would break something via other changes but anyway am not get into that debate of what is the value of testing... 

> I don't see ZK integration tests in druid that verifies ZK config, likes of acl, paths etc., against the ZK server.

Maybe it is true for ACLs but there is a good amount of testing for zookeeper stuff... 

Anyway, Overall looks pretty simple change so LGTM.

File: server/src/main/java/io/druid/curator/CuratorModule.java (hunk 1 above)
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: Please use `StringUtils.format()` instead.

File: server/src/main/java/io/druid/curator/CuratorConfig.java [hunk 2]
```diff
@@ -41,6 +41,15 @@
   @JsonProperty("acl")
   private boolean enableAcl = false;
 
+  @JsonProperty("user")
+  private String zkUser;
+
+  @JsonProperty("pwd")
+  private String zkPwd;
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorConfig.java]: Please use `PasswordProvider` instead of a plain string configuration.

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 3]
```diff
@@ -96,8 +103,7 @@ public void stop()
             log.info("Stopping Curator");
             framework.close();
           }
-        }
-    );
+        });
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: Why did you change this? The line should be broken before `)`.

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 4]
```diff
@@ -114,13 +120,11 @@ public EnsembleProvider makeEnsembleProvider(CuratorConfig config, ExhibitorConf
         new Exhibitors(
             exConfig.getHosts(),
             exConfig.getRestPort(),
-            newBackupProvider(config.getZkHosts())
-        ),
+            newBackupProvider(config.getZkHosts())),
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: Same here. The line should be broken before `)`.

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 5]
```diff
@@ -114,13 +120,11 @@ public EnsembleProvider makeEnsembleProvider(CuratorConfig config, ExhibitorConf
...
         new DefaultExhibitorRestClient(exConfig.getUseSsl()),
         exConfig.getRestUriPath(),
         exConfig.getPollingMs(),
-        new BoundedExponentialBackoffRetry(BASE_SLEEP_TIME_MS, MAX_SLEEP_TIME_MS, MAX_RETRIES)
-    )
+        new BoundedExponentialBackoffRetry(BASE_SLEEP_TIME_MS, MAX_SLEEP_TIME_MS, MAX_RETRIES))
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: Same here. The line should be broken before `)`.

File: server/src/main/java/io/druid/curator/CuratorConfig.java [hunk 6]
```diff
@@ -82,4 +91,35 @@ public void setEnableAcl(Boolean enableAcl)
     Preconditions.checkNotNull(enableAcl, "enableAcl");
     this.enableAcl = enableAcl;
   }
+
+  public String getZkUser()
+  {
+    return zkUser;
+  }
+
+  public void setZkUser(String zkUser)
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorConfig.java]: I think we don't need `set` methods. Let's remove all of them.

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 7]
```diff
@@ -67,18 +71,21 @@ public void configure(Binder binder)
   @Provides
   @LazySingleton
   public CuratorFramework makeCurator(
-      CuratorConfig config, EnsembleProvider ensembleProvider, Lifecycle lifecycle
-  )
+      CuratorConfig config, EnsembleProvider ensembleProvider, Lifecycle lifecycle)
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: Same here. The line should be broken before `)`.

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 8]
```diff
@@ -67,18 +71,21 @@ public void configure(Binder binder)
...
-                               .ensembleProvider(ensembleProvider)
-                               .sessionTimeoutMs(config.getZkSessionTimeoutMs())
-                               .retryPolicy(new BoundedExponentialBackoffRetry(
-                                   BASE_SLEEP_TIME_MS, MAX_SLEEP_TIME_MS, MAX_RETRIES))
-                               .compressionProvider(new PotentiallyGzippedCompressionProvider(config.getEnableCompression()))
-                               .aclProvider(config.getEnableAcl() ? new SecuredACLProvider() : new DefaultACLProvider())
-                               .build();
+    final Builder builder = CuratorFrameworkFactory.builder();
+    if (null != config.getZkUser() && null != config.getZkPwd()) {
+      builder.authorization(config.getAuthScheme(),
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: In Druid's coding convention, this should be 

```
      builder.authorization(
          config.getAuthScheme(),
          (config.getZkUser() + ":" + config.getZkPwd()).getBytes(StandardCharsets.UTF_8)
      );
```

File: server/src/main/java/io/druid/curator/CuratorModule.java [hunk 9]
```diff
@@ -67,18 +71,21 @@ public void configure(Binder binder)
...
+          (config.getZkUser() + ":" + config.getZkPwd()).getBytes(StandardCharsets.UTF_8));
+    }
+    final CuratorFramework framework = builder
+        .ensembleProvider(ensembleProvider)
+        .sessionTimeoutMs(config.getZkSessionTimeoutMs())
+        .retryPolicy(new BoundedExponentialBackoffRetry(
+            BASE_SLEEP_TIME_MS, MAX_SLEEP_TIME_MS, MAX_RETRIES))
```
Comment (jihoonson) [inline server/src/main/java/io/druid/curator/CuratorModule.java]: Don't have to break the line.

**Comment by varaga** [comment]
@jihoonson - I'm using eclipse code formatting file as listed in the contributing page. Has that been outdated by any chance?

**Comment by jihoonson** [comment]
@varaga I don't think so, but it doesn't check everything unfortunately. You should additionally check the code style by comparing with existing codes.

**Comment by chakravarthyvp** [comment]
Thanks for your comments @jihoonson  @b-slim 
I squashed commits into this PR with the review comments implemented.
The build seems to fail and I don't think it is do with these changes. Not sure what's wrong

File: docs/content/configuration/index.md [hunk 10]
```diff
@@ -42,6 +42,9 @@ We recommend just setting the base ZK path and the ZK service host, but all ZK p
 |--------|-----------|-------|
 |`druid.zk.paths.base`|Base Zookeeper path.|`/druid`|
 |`druid.zk.service.host`|The ZooKeeper hosts to connect to. This is a REQUIRED property and therefore a host address must be supplied.|none|
+|`druid.zk.service.user`|The username to authenticate with ZooKeeper. This is an optional property.|none|
+|`druid.zk.service.pwd`|The password to authenticate with ZooKeeper. This is an optional property.|none|
```
Comment (jihoonson) [inline docs/content/configuration/index.md]: Would you please add an explanation that there are two options for configuring password and a link to `PasswordProvider`? http://druid.io/docs/latest/development/extensions-core/mysql.html#configuration can be a good reference.

**Comment by jihoonson** [comment]
@varaga thanks. I left one more comment for doc. Also, would you please fill CLA (http://druid.io/community/cla.html)?

**Comment by varaga** [comment]
@jihoonson  - Done with CLA & the comment !

**Comment by jihoonson** [comment]
We have a regular release cycle for major releases per 3 months. Our next feature freeze is supposed to be around July 7th. After that, we will discuss our next major release (0.13.0), build and test maybe several RCs, and finally make an official release.

File: docs/content/configuration/index.md [hunk 11]
```diff
@@ -42,6 +42,9 @@ We recommend just setting the base ZK path and the ZK service host, but all ZK p
 |--------|-----------|-------|
 |`druid.zk.paths.base`|Base Zookeeper path.|`/druid`|
 |`druid.zk.service.host`|The ZooKeeper hosts to connect to. This is a REQUIRED property and therefore a host address must be supplied.|none|
+|`druid.zk.service.user`|The username to authenticate with ZooKeeper. This is an optional property.|none|
+|`druid.zk.service.pwd`|The [Password Provider](../operations/password-provider.html) or the string password to authenticate with ZooKeeper. This is an optional property.|none|
+|`druid.zk.service.authScheme`|The authentication scheme as supported and honored by ZooKeeper.|digest|
```
Comment (jon-wei) [inline docs/content/configuration/index.md]: Are there any ZK auth schemes other than "digest" that have "user:password" auth data format? If not, let's note here that only "digest" is supported.

Comment (varaga) [reply docs/content/configuration/index.md]: you are right. It'`digest` and `kerebros` are the only authentication supported. I'll update the doc as `digest` is the only supported.

**Comment by varaga** [comment]
@jon-wei  - Modified the doc as suggested and squashed the commits into one.

**Comment by jon-wei** [comment]
@varaga thanks for the contribution!

**Comment by jon-wei** [comment]
@varaga It would most likely be included our first Apache incubator release, 0.13.0, in the meantime if you need support for digest auth I'd recommend seeing if setting up a JAAS client config as described here (https://cwiki.apache.org/confluence/display/ZOOKEEPER/Client-Server+mutual+authentication) works for you
